_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,231 @@
+        }
+      ]
+    }
+  ]
+}
+```
+
+This structure becomes the following `addressBook.proto` file:
+
+```js
+syntax = "proto3";
+
+package tutorial;
+
+message Person {
+  optional string name = 1;
+
+  enum PhoneType {
+    MOBILE = 0;
+    HOME = 1;
+    WORK = 2;
+  }
+
+  message PhoneNumber {
+    optional string phoneNumber = 1;
+    optional PhoneType phoneType = 2;
+  }
+
+  repeated PhoneNumber phones = 2;
+}
+
+message AddressBook {
+  repeated Person people = 1;
+}
+```
+
+## Compile the TypeScript You Will Use in Your Code
+This is one of the bits that didn’t quite make sense when I was first trying to learn Protobufs.  There is some magic that's needed to turn the `.proto` files into language specific code you can use within your project.
+
+- For this you need the `protoc` compiler. Download and install the [prebuilt binary](https://github.com/protocolbuffers/protobuf/releases)
+  - On my Mac I downloaded `protoc-x.xx.x-osx-x86_64.zip`, unzipped and copied the `protoc` file in to my `$PATH` at `/usr/local/bin/`.
+
+However, this only generates code for [some languages](https://github.com/protocolbuffers/protobuf#protobuf-runtime-installation). From that list we're interested in JavaScript. The bit that isn’t included with `protoc` is generating the Typescript types. For that you need a plugin for `protoc` like the [`ts-proto`](https://www.npmjs.com/package/ts-proto) npm package.
+
+Let's start building.  Start a TypeScript project and create an `addressBook.proto` file as above in the same project. Then run:
+
+- `npm install ts-protoc-gen`
+
+Next compile your `.proto` file into a `.ts` file for use within your project by running the following CLI command:
+
+```
+protoc \
+    --plugin="./node_modules/.bin/protoc-gen-ts_proto" \
+    --ts_proto_opt=esModuleInterop=true \
+    --ts_proto_out="./src/generated" \
+    src/proto/addressBook.proto
+```
+
+This is the point where I really like TypeScript. If you run `protoc` without the `ts-proto` plugin and look at the generated code there are 760 lines of stuff in the `generated/src/proto/addressBook.js` file which honestly is not that easy to read or understand how to use. But if you look at the `generated/src/proto/addressBook.ts` file, there are only 250 lines and it provides you with the type interfaces and commands you are going to need to create and use your messages and Protobufs.
+
+_*This is where I fell over on my first attempt, I did not use a plugin like `ts-proto` to generate the types and only had the JS output. I couldn't easily enough understand what was going on, got lost and then distracted by other shiny stuff.*_
+
+## Create a Message
+_*When you read the docs and get presented with ["The API is not well-documented yet"](https://github.com/protocolbuffers/protobuf/tree/master/js#api) you can be forgiven for wanting to cry and giving up. This is where I failed on my second attempt and this is where the Typescript types come to the rescue.*_
```

Comment:
Does this belong here?  Or lower down.

You haven't introduced anything between your two narrative items which makes this jarring.

Not using `ts-node` meant you failed on your first attempt.
What was different between attempt 1 and attempt 2?

Or perhaps:

What caused you to fail on attempt 2?
And what did you learn (and are sharing here) that enabled you to overcome that barrier?

A more natural flow would be:
* here's a thing (ts-node)
* This is why I failed on attempt 1 (but succeeded on this attempt)
* here's another thing
* This is why I failed on attempt 2 (but succeeded on this attempt)

Reply:
Went round the houses on this.  When I first wrote the post I started with the chronological process going through the two different plugins but it ended up way too long and not helpful for a reader looking for an answer quickly.  As a result, I created the final 'how to do it another way' section as optional reading for the second plug in. 

I also had the thought about flow regarding the comment on line 125, will work on it.....

Reply:
Moved to line 125 to final section